atc/exec/build/repository.go

```diff
@@ -21,37 +21,31 @@ type ArtifactName string
 // execution.
 //
 type Repository struct {
-	repo  map[ArtifactName]runtime.Artifact
+	repo  map[ArtifactName]runtime.Volume
 	repoL sync.RWMutex
 
 	parent *Repository
 }
 
-// NewArtifactRepository constructs a new repository.
+// NewRepository constructs a new repository.
 func NewRepository() *Repository {
 	return &Repository{
-		repo: make(map[ArtifactName]runtime.Artifact),
+		repo: make(map[ArtifactName]runtime.Volume),
 	}
 }
 
-//counterfeiter:generate . RegisterableArtifact
-// A RegisterableArtifact is an Artifact which can be added to the registry
-type RegisterableArtifact interface {
-	runtime.Artifact
-}
-
-// RegisterArtifact inserts an RegisterableArtifact into the map under the given
+// RegisterArtifact inserts an artifact Volume into the map under the given
 // ArtifactName. Producers of artifacts, e.g. the Get step and the Task step,
 // will call this after they've successfully produced their artifact(s).
-func (repo *Repository) RegisterArtifact(name ArtifactName, artifact RegisterableArtifact) {
+func (repo *Repository) RegisterArtifact(name ArtifactName, volume runtime.Volume) {
```

Comment:
I might be personally used to seeing `Artifact`, but I wonder if it's worth it to rename it to `Volume` everywhere. I think they're used interchangeably (with the exception of `api/artifactserver`) and might confuse people

Reply:
Yeah fair point, it's a little awkward right now since it's only half committing to calling things volumes. It's weird having code like:

```go
volume, found := repository.ArtifactFor("foo")
```

On the flip side, we often describe inputs/outputs in a build as artifacts, so it'd be nice to keep that terminology.

WDYT about re-introducing a `runtime.Artifact` interface - only now, it's just the required subset of required methods (in particular, just `StreamOut`). Then, `runtime.Input` (and `runtime.ImageSpec`) could take a `runtime.Artifact` instead of a `runtime.Volume` - and runtime implementations could just special case on `Artifacts` that *also* happen to be volumes on the same worker.

(the motivation for something like this came from a discussion on discord where someone was looking for a feature to mount a volume to a `task` based on the contents of a variable - https://discord.com/channels/219899946617274369/413770960089382922/832341657202458744)

Reply:
So the abstraction would be: an `Artifact` is an output of a step, you can grab the bits from it (`StreamOut`), but shouldn't be able to modify it. In this case an `Artifact` is passed from step to step and is purely a Concourse concept (i.e. it doesn't need to be generated from a container-based step). A `Volume` is the interface between `Artifact` and container-based steps, it has all the bells and whistles to take an `Artifact`, and convert it to a form that can be passed into a container runtime like garden.

I like it.

Reply:
Just to make sure we're on the same page, my thinking was that a `Volume` would be a specific type of `Artifact` that is/was mounted to a container-based step. `Volume` aren't necessarily the *only* type of `Artifact`, though.

Any `Artifact` can be used as an input (or image), and it's up to the individual runtimes to convert `Artifacts` to mountable `Volumes`, if necessary. For the `gardenruntime`, this either mean streaming the contents of the `Artifact` into a new empty `Volume`, or creating a COW `Volume` if the `Artifact` was a `Volume` on the same worker.

When you create a container, you get back the list of `Volumes` that were mounted - these can then be registered to the build repository as `Artifacts`. So, when the `exec` package fetches an `Artifact` from the build repo, it has no idea whether it's a `Volume` or not - and it's up to the runtimes to figure out what to do with that (which may involve asking questions like "is this a `Volume`, and if so, what worker is it on?")

Reply:
Take a look at https://github.com/concourse/concourse/pull/6597/commits/e4a4d2b99cb226cc518ba49d5ad7236741d4525f which implements this idea. I'm a bit worried it's premature abstraction, as it adds a *bit* of complexity to the runtime implementation and the `Streamer` (they now need to consider the case where the `Artifact` is not a `Volume`), but I don't think it's too bad. Anyway, happy to revert

Reply:
+1 to most of what you said, the only thing I'm conflicted on is:

> When you create a container, you get back the list of `Volumes` that were mounted - these can then be registered to the build repository as `Artifacts`

Conceptually, `Artifacts` are the _outputs_ of a (previous) step, they should only be known at the end of a container's execution. But the way it's implemented is that it makes the underlying assumption that the volume mounts returned at the beginning of container creation are modified over the course of the container and will hold the output after the container has exited. Maybe `runtime.Container` interface should have a `OutputArtifacts()` that can be called after the process has exited to actually get the `Artifacts` that can be stored in the build repository.

But that's super pedantic and I'm fine with keeping the current implementation.

Reply:
That's a good point. I took a look to remind myself how the `VolumeMounts` (returned by `FindOrCreateContainer`) are used, and it's used for 3 things:

1. Finding the resource volume for the `get` step to register as an artifact *and* to initialize the resource cache
2. Finding the output volumes of a `task` step to register as artifacts
3. Finding the cache volumes of a `task` step to initialize as task caches

For case 2, `OutputArtifacts` makes a lot of sense. For case 1, `OutputArtifacts` would have to still return `runtime.Volumes` (rather than `runtime.Artifacts). For case 3, we'd maybe need a `CacheVolumes` or something along those lines.

Cool if I just leave it as it is for now until we figure out a better approach?